engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc

```diff
@@ -0,0 +1,1264 @@
+  // is an intersection.
+  static std::optional<PinIntersection> ComputeIntersection(UmbraPin* pPin0,
+                                                            UmbraPin* pPin1);
+
+  // Remove the pin at pPin from the linked list of pins because the caller
+  // determined that it should not contribute to the final umbra polygon.
+  // The pointer to the head pin at *pHead will also be adjusted if we've
+  // eliminated the head pin itself and it will be additionally set to
+  // nullptr if that was the last pin in the list.
+  static void RemovePin(UmbraPin* pPin, UmbraPin** pHead);
+
+  // A helper method for resolving pin conflicts, adopted directly from the
+  // associated Skia algorithm.
+  static int ComputeSide(const Point& p0, const Vector2& v, const Point& p);
+
+  // Run through the path calculating the outset vertices for the penumbra
+  // and connecting them to the inset vertices of the umbra and then to
+  // the centroid in a system of triangles with the appropriate alpha values
+  // representing the intensity of the (non-gamma-adjusted) shadow at those
+  // points. The resulting mesh should consist of 2 rings of triangles, an
+  // inner ring connecting the centroid to the umbra polygon, and another
+  // outer ring connecting vertices in the umbra polygon to vertices on the
+  // outer edge of the penumbra.
+  void ComputeMesh();
+
+  // After the umbra_vertices of the pins are accumulated and linked into a
+  // ring using their pPrev/pNext pointers, compute the best surviving umbra
+  // vertex for each pin and set its location and index into the UmbraPin.
+  void PopulateUmbraVertices();
+
+  // Appends a fan of penumbra vertices centered on the path vertex of the
+  // |p_curr_pin| starting from the absolute point |fan_start| and ending
+  // at the absolute point |fan_end|, both of which should be equi-distant
+  // from the path vertex. The index of the vertex at |fan_start| should
+  // already be in the vector of vertices at an index given by |start_index|.
+  uint16_t AppendFan(const UmbraPin* p_curr_pin,
+                     const Point& fan_start,
+                     const Point& fan_end,
+                     uint16_t start_index);
+
+  // Append a vertex and its associated gaussian coefficient to the lists
+  // of vertices and guassians and return their (shared) index.
+  uint16_t AppendVertex(const Point& vertex, Scalar gaussian);
+
+  // Append 3 indices to the indices vector to form a new triangle in the mesh.
+  void AddTriangle(uint16_t v0, uint16_t v1, uint16_t v2);
+};
+
+PolygonInfo::PolygonInfo(const impeller::PathSource& source,
+                         const Matrix& matrix,
+                         Scalar occluder_height,
+                         const Tessellator::Trigs& trigs)
+    : inverted_matrix_(matrix.Invert()),
+      occluder_height_(occluder_height),
+      trigs_(trigs) {
+  if (!matrix.IsInvertible()) {
+    Invalidate();
+    return;
+  }
```

Comment:
Instead of creating invalid objects, please use absl::StatusOr with a factory function.